File: 2-Stack/4-daily_temperatures.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

vector<int> dailyTemperatures(vector<int>& temperatures) {
    int len = temperatures.size();
    vector<int> days(len, 0);
    stack<int> temps;

    for (int i = 0; i < len; i++) {
        while (!temps.empty() && temperatures[i] > temperatures[temps.top()]) {
            int prev = temps.top();
            temps.pop();
            days[prev] = i - prev;
        }
        temps.push(i);
    }
    return days;
}

int main() {
    
}

File: PYTHON/2-Stack/4-daily_temperatures.py
from typing import List

def dailyTemperatures(temperatures: List[int]) -> List[int]:
    days = [0] * len(temperatures)
    temps = []

    for i in range(len(temperatures)):
        while temps and temperatures[i] > temperatures[temps[-1]]:
            prev = temps.pop()
            days[prev] = i - prev
        temps.append(i)
    
    return days

if __name__ == "__main__":
    temperatures = [30, 38, 30, 36, 35, 40, 28]
    print(dailyTemperatures(temperatures))